src/api/logs/logger_provider.zig

```diff
@@ -1,144 +1,408 @@
+/// see: https://opentelemetry.io/docs/specs/otel/logs/sdk/#logrecordprocessor
+pub const ReadWriteLogRecord = struct {
+    timestamp: ?u64,
+    observed_timestamp: u64,
+    trace_id: ?[16]u8,
+    span_id: ?[8]u8,
+    severity_number: ?u8,
+    severity_text: ?[]const u8,
+    body: ?[]const u8,
+    attributes: std.ArrayListUnmanaged(Attribute),
+    resource: ?[]const Attribute,
     scope: InstrumentationScope,
 
     const Self = @This();
 
-    pub fn init(allocator: std.mem.Allocator, scope: InstrumentationScope) !*Self {
-        const logger = try allocator.create(Self);
-
-        logger.* = Self{
-            .allocator = allocator,
+    pub fn init(_: std.mem.Allocator, scope: InstrumentationScope) Self {
+        return Self{
+            .timestamp = null,
+            .observed_timestamp = @intCast(std.time.nanoTimestamp()),
+            .trace_id = null,
+            .span_id = null,
+            .severity_number = null,
+            .severity_text = null,
+            .body = null,
+            .attributes = std.ArrayListUnmanaged(Attribute){},
+            .resource = null,
             .scope = scope,
         };
+    }
 
-        return logger;
+    pub fn setAttribute(self: *Self, allocator: std.mem.Allocator, attribute: Attribute) !void {
+        try self.attributes.append(allocator, attribute);
     }
 
-    pub fn deinit(self: *Self) void {
-        self.allocator.destroy(self);
+    pub fn deinit(self: *Self, allocator: std.mem.Allocator) void {
+        self.attributes.deinit(allocator);
+    }
+
+    /// Convert to immutable ReadableLogRecord for export
+    pub fn toReadable(self: *const Self, allocator: std.mem.Allocator) !ReadableLogRecord {
+        const attrs = try allocator.alloc(Attribute, self.attributes.items.len);
+        @memcpy(attrs, self.attributes.items);
+
+        return ReadableLogRecord{
+            .timestamp = self.timestamp,
+            .observed_timestamp = self.observed_timestamp,
+            .trace_id = self.trace_id,
+            .span_id = self.span_id,
+            .severity_number = self.severity_number,
+            .severity_text = self.severity_text,
+            .body = self.body,
+            .attributes = attrs,
+            .resource = self.resource,
+            .scope = self.scope,
+        };
+    }
+};
+
+/// ReadableLogRecord is an immutable log record passed to exporters.
+/// see: https://opentelemetry.io/docs/specs/otel/logs/sdk/#logrecordexporter
+pub const ReadableLogRecord = struct {
+    timestamp: ?u64,
+    observed_timestamp: u64,
+    trace_id: ?[16]u8,
+    span_id: ?[8]u8,
+    severity_number: ?u8,
+    severity_text: ?[]const u8,
+    body: ?[]const u8,
+    attributes: []const Attribute,
+    resource: ?[]const Attribute,
+    scope: InstrumentationScope,
+
+    const Self = @This();
+
+    pub fn deinit(self: Self, allocator: std.mem.Allocator) void {
+        allocator.free(self.attributes);
```

Comment:
When the resource is populated, we also need to free the backing array.

```suggestion
        allocator.free(self.attributes);
        if(self.resource) |r| allocator.free(r);
```

I would suggest adding a test that covers it to validate the assumption that there might be memory leak.

Reply:
I saw the test below "Log records inherit resource from provider" after commenting: we need to decide what to do with the ownership of resource attributes.

It works in the test below because the resource is owned by the provider, so I think to avoid mistakes from users we should duplicate the array at L64 and free it on `deinit`.

The risk otherwise is that a resource backing array is freed wile the record is being used.